src/main/drivers/dshot_bitbang_decode.c

```diff
@@ -0,0 +1,133 @@
+#include <stdint.h>
+#include <string.h>
+#include <stdio.h>
+
+#include "platform.h"
+
+#if defined(USE_DSHOT_TELEMETRY)
+
+#include "common/maths.h"
+#include "common/utils.h"
+#include "drivers/dshot.h"
+#include "drivers/dshot_bitbang_decode.h"
+
+#define MIN_VALID_BBSAMPLES ((21 - 2) * 3)
+#define MAX_VALID_BBSAMPLES ((21 + 2) * 3)
+
+#ifdef DEBUG_BBDECODE
+uint16_t bbBuffer[134];
+#endif
+
+FAST_CODE uint32_t decode_bb( uint16_t buffer[], uint32_t count, uint32_t mask)
+{
+//    printf("start\n");
+#ifdef DEBUG_BBDECODE
+    uint32_t sequence[MAX_GCR_EDGES];
+    memset(sequence, 0, sizeof(sequence));
+    int sequenceIndex = 0;
+#endif
+
+    uint16_t lastValue = 0;
+    uint32_t value = 0;
+
+    uint16_t* p = buffer;
+    uint16_t* endP = p + count - MIN_VALID_BBSAMPLES;
+    while (p < endP) {
+        if (__builtin_expect(!(*p++ & mask), 0) ||
+            __builtin_expect(!(*p++ & mask), 0) ||
+            __builtin_expect(!(*p++ & mask), 0) ||
+            __builtin_expect(!(*p++ & mask), 0)) {
+            break;
+        }
+    }
+
+    if(*p & mask) {
```

Comment:
:space_invader:

Reply:
Some documentation in this section would be really nice. The code is very obscure and not many people are going to understand it later.

Reply:
I've documented the whole function now.